cmd/execute.cpp

```diff
@@ -115,59 +161,84 @@ int main(int argc, char* argv[]) {
 
     CLI11_PARSE(app, argc, argv);
 
-    uint64_t batch_size{batch_mib * 1024 * 1024};
-
-    std::clog << "Starting block execution. DB: " << db_path << std::endl;
+    // Check data.mdb exists in provided directory
+    boost::filesystem::path db_file{boost::filesystem::path(db_path) / boost::filesystem::path("data.mdb")};
+    if (!boost::filesystem::exists(db_file)) {
+        std::clog << "Can't find a valid TG data file in " << db_path << std::endl;
+        return -2;
+    }
 
-    MDB_env* env{nullptr};
-    check_lmdb_err(mdb_env_create(&env));
-    auto cleanup{gsl::finally([env] { mdb_env_close(env); })};
+    // Check provided map size is valid
+    auto map_size{parse_size(map_size_str)};
+    if (!map_size.has_value()) {
+        std::clog << "Invalid --lmdb.mapSize value provided : " << map_size_str << std::endl;
+        return -2;
+    }
 
-    check_lmdb_err(mdb_env_set_mapsize(env, 1ull << 40));
-    check_lmdb_err(mdb_env_set_maxdbs(env, 128));
-    check_lmdb_err(mdb_env_open(env, db_path.c_str(), MDB_NOTLS | MDB_NORDAHEAD | MDB_NOSYNC, 0644));
+    std::clog << "Starting block execution. DB: " << db_file << std::endl;
+
+    silkworm::lmdb::DatabaseConfig db_config{db_path, *map_size};
+    db_config.set_readonly(false);
+    std::shared_ptr<silkworm::lmdb::Environment> env{nullptr};
+    std::unique_ptr<silkworm::lmdb::Transaction> txn{nullptr};
+
+    try
+    {
+        env = silkworm::lmdb::get_env(db_config);
+        txn = env->begin_rw_transaction();
+        bool write_receipts{storage_mode_has_write_receipts(txn)};
+        if (write_receipts && (!migration_happened(txn, "receipts_cbor_encode") ||
+                               !migration_happened(txn, "receipts_store_logs_separately"))) {
+            throw std::logic_error("Legacy stored receipts are not supported");
```

Comment:
Why throw exception and then catch it later? What was wrong with the immediate return?

Reply:
Nothing wrong particularly.
As a mindset I always try to keep returns from main to the very bottom